packages/nextjs/components/BatchStatusIcon.tsx

```diff
@@ -0,0 +1,96 @@
+import { FaCheckCircle, FaTimesCircle } from "react-icons/fa";
+import { useScaffoldReadContract } from "~~/hooks/scaffold-eth";
+
+export const BatchStatusIcon = ({ userAddress }: { userAddress: string | undefined }): React.ReactNode => {
+  const { data: allowList } = useScaffoldReadContract({
+    contractName: "BatchRegistry",
+    functionName: "allowList",
+    args: [userAddress],
+  });
+
+  const { data: yourContractAddress } = useScaffoldReadContract({
+    contractName: "BatchRegistry",
+    functionName: "yourContractAddress",
+    args: [userAddress],
+  });
+
+  function isInBatch13() {
+    return allowList;
+  }
+  function isCheckedIn() {
+    return yourContractAddress == "0x0000000000000000000000000000000000000000" ? false : true;
+  }
+
+  /**
+   * Batch Info header icons
+   */
+  interface Batch13Icon {
+    className?: string; // Additional classes can be added
+  }
+
+  const Batch13Icon: React.FC<Batch13Icon> = ({ className }) => {
+    return (
+      <div
+        className={`batch-13-icon flex items-center justify-center dark:text-white font-bold px-2 rounded-full shadow-lg ${className}`}
+        style={{ width: 30, height: 30, fontSize: 30 * 0.4 }}
+      >
+        13
+      </div>
+    );
+  };
+
+  interface CheckedInIcon {
+    className?: string; // Additional classes can be added
+    checkedIn?: boolean; // Used to display if connected address is checked into Batch 13 contract or not
+  }
+
+  const CheckedInIcon: React.FC<CheckedInIcon> = ({ className = "", checkedIn }) => {
+    return (
+      <div
+        className={`batch-13-icon flex items-center justify-center text-black dark:text-white font-bold px-2 rounded-full shadow-lg ${checkedIn ? "dark:bg-green-800 bg-green-500" : "dark:bg-red-800 bg-red-500"}${className}`}
+        style={{ width: 30, height: 30, fontSize: 30 * 0.4 }}
+      >
+        {checkedIn ? <FaCheckCircle /> : <FaTimesCircle />}
+      </div>
+    );
+  };
+
+  /**
+   * State Machine
+   */
+  const userState = () => {
+    if (!isInBatch13()) return "NOT_IN_BATCH";
+    if (!isCheckedIn()) return "IN_BATCH_NOT_CHECKED_IN";
+    return "IN_BATCH_CHECKED_IN";
+  };
```

Comment:
We can apply the same thing I suggested in a previous comment, but in this case, we can use the `useMemo` hook so React will only calculate the value again if its dependencies change. I suggest taking a look at the documentation if this is new for you.

```suggestion
  const userState = useMemo(() => {
    if (!isInBatch13) return "NOT_IN_BATCH";
    if (!isCheckedIn) return "IN_BATCH_NOT_CHECKED_IN";
    return "IN_BATCH_CHECKED_IN";
  }, [isInBatch13, isCheckedIn]);
```
After this change, you can simply use `userState` instead of `userState()`.

There are some situations where you will need a function, like when you need to pass an argument, but this is not the case.

Reply:
Oh, and I just noticed, the name `userState` is not very good. It's too similar to `useState`, which is a very common React hook.

Reply:
ok renamed it to batchState instead and I did consider using useState initially but saw on first render it was giving a null value so didn't end up using it. Forgot all about useMemo, thank you very much for this better way of doing it.